src/lib/PageTracker.js

```diff
@@ -0,0 +1,310 @@
+const CONTENT_INSIGHTS_IDLE_TIMER = 30000;
+const CONTENT_INSIGHTS_ACTIVE_USER_THROTTLE = 500;
+
+class PageTracker {
+    /** @property {Api} - PageTracker Api instance used for XHR calls  */
+    api = new Api();
+
+    /** @property {number} -  */
+    currentPage = null;
+
+    /** @property {Object} - Current file being previewed */
+    file = {};
+
+    /** @property {string} -  */
+    fileType = 'document';
+
+    /** @property {number} -  */
+    fileLength = null;
+
+    /** @property {number} -  */
+    idleTimerTimeout = null;
+
+    /** @property {number} -  */
+    reportingInterval = null;
+
+    /** @property {boolean} -  */
+    reportingStarted = false;
+
+    /** @property {object} -  */
+    requestOptions = {
+        apiHost: '',
+        headers: {},
+    };
+
+    /** @property {string} -  */
+    sessionId = uuidv4();
+
+    /** @property {Date} -  */
+    startTime = null;
+
+    /** @property {Object} - Reference to content insights object */
+    contentInsightsProps = {
+        thumbnailImageCache: new BoundedCache(),
+    };
+
+    contentInsightsOptions = {
+        contentInsightsConfig: {
+            isActive: false,
+        },
+    };
+
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+
+    /**
+     * [constructor]
+     *
+     * @return {PageTracker} PageTracker instance
+     */
+    constructor() {
+        this.activeUserHandler = this.activeUserHandler.bind(this);
+        this.inactiveUserHandler = this.inactiveUserHandler.bind(this);
+        this.throttledActiveUserHandler = this.activeUserHandler().bind(this);
+    }
+
+    /**
+     * Get the session id
+     *
+     * @public
+     * @return {string}
+     */
+    getSessionId() {
+        return this.sessionId;
+    }
+
+    /**
+     * Call the interval starting function and set the idle timer
+     *
+     * @private
+     * @param {Object} data - Data Object to set the current page and file length
+     * @return {void}
+     */
+    initContentInsightsLog(data) {
+        if (data.currentPage) {
+            this.currentPage = data.currentPage;
+            this.fileLength = data.numPages;
+        }
+        this.setIdleTimer();
+        this.reportingStarted = true;
+        this.startContentInsightsInterval();
+    }
+
+    isContentInsightsActive() {
+        return this.contentInsightsOptions.contentInsightsConfig.isActive;
+    }
+
+    /**
+     * Start the content insights interval, to report data each number of seconds defined on a Const
+     *
+     * @public
+     * @return {void}
+     */
+    startContentInsightsInterval() {
+        if (!this.reportingStarted) {
+            return;
+        }
+        if (this.reportingInterval) {
+            this.stopContentInsightsInterval();
+        }
+        this.startTime = Date.now();
+
+        this.reportingInterval = setInterval(() => {
```

Comment:
With our experience, the page unload events weren't always reliable to send the data back to an API. We tried looking up also with websockets but they were failing too. In the case of sendBeacon, a year ago we read some articles that this new Api is not recommended because sometimes it fails to fire the events, so we stick with the setInterval method. In case we will stuck in a weird situation where executions start backing up, I think it won't be a big Issue because we don't necessary rely on the events order.